packages/server/graphql/private/mutations/processRecurrence.ts

```diff
@@ -0,0 +1,118 @@
+
+const startRecurringTeamPrompt = async (
+  meetingSeries: MeetingSeries,
+  startTime: Date,
+  dataLoader,
+  r,
+  subOptions
+) => {
+  const {teamId, facilitatorId} = meetingSeries
+
+  // AUTH
+  const unpaidError = await isStartMeetingLocked(teamId, dataLoader)
+  if (unpaidError) return standardError(new Error(unpaidError), {userId: facilitatorId})
+
+  const meeting = await createTeamPrompt(teamId, facilitatorId, r, dataLoader)
+
+  const formattedDate = new Date().toLocaleDateString('en-US', {
+    month: 'short',
+    day: 'numeric'
+  })
+
+  meeting.name = `${meetingSeries.title} - ${formattedDate}`
+  meeting.scheduledEndTime = new Date(startTime.getTime() + ms(`${meetingSeries.duration}m`))
+  meeting.meetingSeriesId = meetingSeries.id
+
+  await r.table('NewMeeting').insert(meeting).run()
+
+  IntegrationNotifier.startMeeting(dataLoader, meeting.id, teamId)
+  analytics.meetingStarted(facilitatorId, meeting, undefined, true)
+  const data = {teamId, meetingId: meeting.id}
+  publish(SubscriptionChannel.TEAM, teamId, 'StartTeamPromptSuccess', data, subOptions)
+}
+
+const processRecurrence: MutationResolvers['processRecurrence'] = async (
+  _source,
+  {},
+  {authToken, dataLoader, socketId: mutatorId}
+) => {
+  const r = await getRethink()
+  const now = new Date()
+  const operationId = dataLoader.share()
+  const subOptions = {mutatorId, operationId}
+
+  // VALIDATION
+
+  // RESOLUTION
+  // Find any meetings with a scheduledEndTime before now, and close them
+  const teamPromptMeetingsToEnd = (await r
+    .table('NewMeeting')
+    .filter({meetingType: 'teamPrompt'})
+    .filter((row) => row.hasFields('endedAt').not())
+    .filter((row) => row('scheduledEndTime').lt(now))
+    .run()) as MeetingTeamPrompt[]
+
+  let meetingsEnded = 0
+  await Promise.all(
+    teamPromptMeetingsToEnd.map(async (meeting) => {
+      const res = await endTeamPrompt({meeting, now, dataLoader, r, subOptions})
+      if (!('error' in res)) meetingsEnded++
+    })
+  )
+
+  let meetingsStarted = 0
+
+  // For each active meeting series, get the meeting start times (according to rrule) after the most
+  // recent meeting start time and before now.
+  const activeMeetingSeries = await getActiveMeetingSeries()
+  for (const meetingSeries of activeMeetingSeries) {
+    const lastMeeting = await r
+      .table('NewMeeting')
+      .filter({meetingType: 'teamPrompt', meetingSeriesId: meetingSeries.id})
+      .orderBy(r.desc('createdAt'))
+      .nth(0)
+      .run()
+
+    // For meetings that should still be active, start the meeting and set its end time.
+    // Any subscriptions are handled by the shared meeting start code
+    const rrule = RRule.fromString(meetingSeries.recurrenceRule)
+
+    // Only get meetings that should currently be active, i.e. meetings that should have started
+    // within the last 24 hours, started after the last meeting in the series, and started before
+    // 'now'.
+    const fromDate = new Date(
```

Comment:
Doing these updates as scheduled jobs would make meeting start times more exact, but it would also introduce more moving parts + complexity, so I have a strong preference for keeping all updates contained within this mutation for now, and introducing more precise scheduling if/when customer and product needs require it.  This is also what was agreed upon in the [spec](https://github.com/ParabolInc/parabol/issues/6888)

Reply:
That's partially my bad - the spec mentions how this is called in the overview but nowhere else. This mutation will be called periodically (once every ~5 minutes initially) by the chronos cron scheduler, so the actual start time should always be within +5 minutes of the scheduled start time regardless.

Reply:
gotcha. I'm hesitant to introduce something that has a 5 min latency because everything else in our app (and all other modern apps) is real time. If the start time is early, it's no big deal, the only way they'd know is the timestamp on the slack message. If the facilitator starts the meeting at the time they want everyone to hop in, that might make them look bad as everyone sits around waiting for it to open, but like you say, future tension.